packages/playwright-core/src/server/dispatchers/webSocketRouteDispatcher.ts

```diff
@@ -0,0 +1,150 @@
+    const context = target instanceof Page ? target.context() : target;
+    if (!(context as any)[kBindingInstalledSymbol]) {
+      (context as any)[kBindingInstalledSymbol] = true;
+
+      await context.exposeBinding('__pwWebSocketBinding', false, (source, payload: ws.BindingPayload) => {
+        if (payload.type === 'onCreate') {
+          const pageDispatcher = PageDispatcher.fromNullable(contextDispatcher, source.page);
+          let scope: PageDispatcher | BrowserContextDispatcher | undefined;
+          if (pageDispatcher && matchesPattern(pageDispatcher, context._options.baseURL, payload.url))
+            scope = pageDispatcher;
+          else if (matchesPattern(contextDispatcher, context._options.baseURL, payload.url))
+            scope = contextDispatcher;
+          if (scope) {
+            new WebSocketRouteDispatcher(scope, payload.id, payload.url, source.frame);
+          } else {
+            const request: ws.ConnectRequest = { id: payload.id, type: 'connect', fuse: true };
+            source.frame.evaluateExpression(`globalThis.__pwWebSocketDispatch(${JSON.stringify(request)})`).catch(() => {});
+          }
+          return;
+        }
+
+        const dispatcher = WebSocketRouteDispatcher._idToDispatcher.get(payload.id);
+        if (payload.type === 'onMessageFromPage')
+          dispatcher?._dispatchEvent('messageFromPage', { message: payload.data.data, isBase64: payload.data.isBase64 });
+        if (payload.type === 'onMessageFromServer')
+          dispatcher?._dispatchEvent('messageFromServer', { message: payload.data.data, isBase64: payload.data.isBase64 });
+        if (payload.type === 'onClose')
+          dispatcher?._onClose();
+      });
+    }
+
+    if (!(target as any)[kInitScriptInstalledSymbol]) {
+      (target as any)[kInitScriptInstalledSymbol] = true;
+      await target.addInitScript(`
+        (() => {
+          const module = {};
+          ${webSocketMockSource.source}
+          (module.exports.inject())(globalThis);
+        })();
+      `);
+    }
+  }
+
+  async connect(params: channels.WebSocketRouteConnectParams) {
+    await this._evaluateAPIRequest({ id: this._id, type: 'connect' });
+  }
+
+  async ensureOpened(params: channels.WebSocketRouteEnsureOpenedParams) {
+    await this._evaluateAPIRequest({ id: this._id, type: 'ensureOpened' });
+  }
+
+  async sendToPage(params: channels.WebSocketRouteSendToPageParams) {
+    await this._evaluateAPIRequest({ id: this._id, type: 'sendToPage', data: { data: params.message, isBase64: params.isBase64 } });
+  }
+
+  async sendToServer(params: channels.WebSocketRouteSendToServerParams) {
+    await this._evaluateAPIRequest({ id: this._id, type: 'sendToServer', data: { data: params.message, isBase64: params.isBase64 } });
+  }
+
+  async close(params: channels.WebSocketRouteCloseParams) {
+    await this._evaluateAPIRequest({ id: this._id, type: 'close', code: params.code, reason: params.reason, wasClean: true });
```

Comment:
Note: Customers like to test with / use `wasClean: false` - might be worth adding to our close API here.